File: sample/java/16_glass_sculpture_chaos.java
final class _16_glass_sculpture_chaos {
    private _16_glass_sculpture_chaos() {
    }


    // 16: Sample that ray-traces chaotic rotation of glass sculptures and outputs a GIF.

    public static double clamp01(double v) {
        if (((v) < (0.0))) {
            return 0.0;
        }
        if (((v) > (1.0))) {
            return 1.0;
        }
        return v;
    }

    public static double dot(double ax, double ay, double az, double bx, double by, double bz) {
        return ax * bx + ay * by + az * bz;
    }

    public static double length(double x, double y, double z) {
        return math.sqrt(x * x + y * y + z * z);
    }

    public static Object normalize(double x, double y, double z) {
        double l = length(x, y, z);
        if (((l) < (1e-09))) {
            return new java.util.ArrayList<Object>(java.util.Arrays.asList(0.0, 0.0, 0.0));
        }
        return new java.util.ArrayList<Object>(java.util.Arrays.asList(x / l, y / l, z / l));
    }

    public static Object reflect(double ix, double iy, double iz, double nx, double ny, double nz) {
        double d = dot(ix, iy, iz, nx, ny, nz) * 2.0;
        return new java.util.ArrayList<Object>(java.util.Arrays.asList(ix - d * nx, iy - d * ny, iz - d * nz));
    }

    public static Object refract(double ix, double iy, double iz, double nx, double ny, double nz, double eta) {
        double cosi = (-(dot(ix, iy, iz, nx, ny, nz)));
        double sint2 = eta * eta * (1.0 - cosi * cosi);
        if (((sint2) > (1.0))) {
            return reflect(ix, iy, iz, nx, ny, nz);
        }
        double cost = math.sqrt(1.0 - sint2);
        double k = eta * cosi - cost;
        return new java.util.ArrayList<Object>(java.util.Arrays.asList(eta * ix + k * nx, eta * iy + k * ny, eta * iz + k * nz));
    }

    public static double schlick(double cos_theta, double f0) {
        double m = 1.0 - cos_theta;
        return f0 + (1.0 - f0) * (m * m * m * m * m);
    }

    public static Object sky_color(double dx, double dy, double dz, double tphase) {
        double t = 0.5 * (dy + 1.0);
        double r = 0.06 + 0.2 * t;
        double g = 0.1 + 0.25 * t;
        double b = 0.16 + 0.45 * t;
        double band = 0.5 + 0.5 * math.sin(8.0 * dx + 6.0 * dz + tphase);
        r += 0.08 * band;
        g += 0.05 * band;
        b += 0.12 * band;
        return new java.util.ArrayList<Object>(java.util.Arrays.asList(clamp01(r), clamp01(g), clamp01(b)));
    }

    public static double sphere_intersect(double ox, double oy, double oz, double dx, double dy, double dz, double cx, double cy, double cz, double radius) {
        double lx = ox - cx;
        double ly = oy - cy;
        double lz = oz - cz;
        double b = lx * dx + ly * dy + lz * dz;
        double c = lx * lx + ly * ly + lz * lz - radius * radius;
        double h = b * b - c;
        if (((h) < (0.0))) {
            return (-(1.0));
        }
        double s = math.sqrt(h);
        double t0 = (-(b)) - s;
        if (((t0) > (0.0001))) {
            return t0;
        }
        double t1 = (-(b)) + s;
        if (((t1) > (0.0001))) {
            return t1;
        }
        return (-(1.0));
    }

    public static java.util.ArrayList<Long> palette_332() {
        java.util.ArrayList<Long> p = PyRuntime.__pytra_bytearray(256L * 3L);
        for (long i = 0L; i < 256L; i += 1L) {
            long r = i >> 5L & 7L;
            long g = i >> 2L & 7L;
            long b = i & 3L;
            p.set((int)((((i * 3L + 0L) < 0L) ? (((long)(p.size())) + (i * 3L + 0L)) : (i * 3L + 0L))), PyRuntime.__pytra_int(((double)(255L * r)) / ((double)(7L))));
            p.set((int)((((i * 3L + 1L) < 0L) ? (((long)(p.size())) + (i * 3L + 1L)) : (i * 3L + 1L))), PyRuntime.__pytra_int(((double)(255L * g)) / ((double)(7L))));
            p.set((int)((((i * 3L + 2L) < 0L) ? (((long)(p.size())) + (i * 3L + 2L)) : (i * 3L + 2L))), PyRuntime.__pytra_int(((double)(255L * b)) / ((double)(3L))));
        }
        return PyRuntime.__pytra_bytearray(p);
    }

    public static long quantize_332(double r, double g, double b) {
        long rr = PyRuntime.__pytra_int(clamp01(r) * 255.0);
        long gg = PyRuntime.__pytra_int(clamp01(g) * 255.0);
        long bb = PyRuntime.__pytra_int(clamp01(b) * 255.0);
        return (rr >> 5L << 5L) + (gg >> 5L << 2L) + (bb >> 6L);
    }

    public static java.util.ArrayList<Long> render_frame(long width, long height, long frame_id, long frames_n) {
        double t = ((double)(frame_id)) / ((double)(frames_n));
        double tphase = 2.0 * math.pi * t;
        double cam_r = 3.0;
        double cam_x = cam_r * math.cos(tphase * 0.9);
        double cam_y = 1.1 + 0.25 * math.sin(tphase * 0.6);
        double cam_z = cam_r * math.sin(tphase * 0.9);
        double look_x = 0.0;
        double look_y = 0.35;
        double look_z = 0.0;
        java.util.ArrayList<Object> __tuple_0 = ((java.util.ArrayList<Object>)(Object)(normalize(look_x - cam_x, look_y - cam_y, look_z - cam_z)));
        double fwd_x = ((Double)(__tuple_0.get(0)));
        double fwd_y = ((Double)(__tuple_0.get(1)));
        double fwd_z = ((Double)(__tuple_0.get(2)));
        java.util.ArrayList<Object> __tuple_1 = ((java.util.ArrayList<Object>)(Object)(normalize(fwd_z, 0.0, (-(fwd_x)))));
        double right_x = ((Double)(__tuple_1.get(0)));
        double right_y = ((Double)(__tuple_1.get(1)));
        double right_z = ((Double)(__tuple_1.get(2)));
        java.util.ArrayList<Object> __tuple_2 = ((java.util.ArrayList<Object>)(Object)(normalize(right_y * fwd_z - right_z * fwd_y, right_z * fwd_x - right_x * fwd_z, right_x * fwd_y - right_y * fwd_x)));
        double up_x = ((Double)(__tuple_2.get(0)));
        double up_y = ((Double)(__tuple_2.get(1)));
        double up_z = ((Double)(__tuple_2.get(2)));
        double s0x = 0.9 * math.cos(1.3 * tphase);
        double s0y = 0.15 + 0.35 * math.sin(1.7 * tphase);
        double s0z = 0.9 * math.sin(1.3 * tphase);
        double s1x = 1.2 * math.cos(1.3 * tphase + 2.094);
        double s1y = 0.1 + 0.4 * math.sin(1.1 * tphase + 0.8);
        double s1z = 1.2 * math.sin(1.3 * tphase + 2.094);
        double s2x = 1.0 * math.cos(1.3 * tphase + 4.188);
        double s2y = 0.2 + 0.3 * math.sin(1.5 * tphase + 1.9);
        double s2z = 1.0 * math.sin(1.3 * tphase + 4.188);
        double lr = 0.35;
        double lx = 2.4 * math.cos(tphase * 1.8);
        double ly = 1.8 + 0.8 * math.sin(tphase * 1.2);
        double lz = 2.4 * math.sin(tphase * 1.8);
        java.util.ArrayList<Long> frame = PyRuntime.__pytra_bytearray(width * height);
        double aspect = ((double)(width)) / ((double)(height));
        double fov = 1.25;
        for (long py = 0L; py < height; py += 1L) {
            long row_base = py * width;
            double sy = 1.0 - 2.0 * (((double)(py)) + 0.5) / ((double)(height));
            for (long px = 0L; px < width; px += 1L) {
                double sx = (2.0 * (((double)(px)) + 0.5) / ((double)(width)) - 1.0) * aspect;
                double rx = fwd_x + fov * (sx * right_x + sy * up_x);
                double ry = fwd_y + fov * (sx * right_y + sy * up_y);
                double rz = fwd_z + fov * (sx * right_z + sy * up_z);
                java.util.ArrayList<Object> __tuple_3 = ((java.util.ArrayList<Object>)(Object)(normalize(rx, ry, rz)));
                double dx = ((Double)(__tuple_3.get(0)));
                double dy = ((Double)(__tuple_3.get(1)));
                double dz = ((Double)(__tuple_3.get(2)));
                double best_t = 1000000000.0;
                long hit_kind = 0L;
                double r = 0.0;
                double g = 0.0;
                double b = 0.0;
                if (((dy) < ((-(1e-06))))) {
                    double tf = ((-(1.2)) - cam_y) / dy;
                    if ((((tf) > (0.0001)) && ((tf) < (best_t)))) {
                        best_t = tf;
                        hit_kind = 1L;
                    }
                }
                double t0 = sphere_intersect(cam_x, cam_y, cam_z, dx, dy, dz, s0x, s0y, s0z, 0.65);
                if ((((t0) > (0.0)) && ((t0) < (best_t)))) {
                    best_t = t0;
                    hit_kind = 2L;
                }
                double t1 = sphere_intersect(cam_x, cam_y, cam_z, dx, dy, dz, s1x, s1y, s1z, 0.72);
                if ((((t1) > (0.0)) && ((t1) < (best_t)))) {
                    best_t = t1;
                    hit_kind = 3L;
                }
                double t2 = sphere_intersect(cam_x, cam_y, cam_z, dx, dy, dz, s2x, s2y, s2z, 0.58);
                if ((((t2) > (0.0)) && ((t2) < (best_t)))) {
                    best_t = t2;
                    hit_kind = 4L;
                }
                double glow = 0.0;
                double hx = 0.0;
                double hz = 0.0;
                double ldx = 0.0;
                double ldy = 0.0;
                double ldz = 0.0;
                double lxv = 0.0;
                double lyv = 0.0;
                double lzv = 0.0;
                double ndotl = 0.0;
                if (((hit_kind) == (0L))) {
                    java.util.ArrayList<Object> __tuple_4 = ((java.util.ArrayList<Object>)(Object)(sky_color(dx, dy, dz, tphase)));
                    r = ((Double)(__tuple_4.get(0)));
                    g = ((Double)(__tuple_4.get(1)));
                    b = ((Double)(__tuple_4.get(2)));
                } else {
                    if (((hit_kind) == (1L))) {
                        hx = cam_x + best_t * dx;
                        hz = cam_z + best_t * dz;
                        long cx_i = PyRuntime.__pytra_int(math.floor(hx * 2.0));
                        long cz_i = PyRuntime.__pytra_int(math.floor(hz * 2.0));
                        long checker = (((((cx_i + cz_i) % 2L) == (0L))) ? (0L) : (1L));
                        double base_r = ((((checker) == (0L))) ? (0.1) : (0.04));
                        double base_g = ((((checker) == (0L))) ? (0.11) : (0.05));
                        double base_b = ((((checker) == (0L))) ? (0.13) : (0.08));
                        lxv = lx - hx;
                        lyv = ly - (-(1.2));
                        lzv = lz - hz;
                        java.util.ArrayList<Object> __tuple_5 = ((java.util.ArrayList<Object>)(Object)(normalize(lxv, lyv, lzv)));
                        ldx = ((Double)(__tuple_5.get(0)));
                        ldy = ((Double)(__tuple_5.get(1)));
                        ldz = ((Double)(__tuple_5.get(2)));
                        ndotl = Math.max(ldy, 0.0);
                        double ldist2 = lxv * lxv + lyv * lyv + lzv * lzv;
                        glow = 8.0 / (1.0 + ldist2);
                        r = base_r + 0.8 * glow + 0.2 * ndotl;
                        g = base_g + 0.5 * glow + 0.18 * ndotl;
                        b = base_b + 1.0 * glow + 0.24 * ndotl;
                    } else {
                        double cx = 0.0;
                        double cy = 0.0;
                        double cz = 0.0;
                        double rad = 1.0;
                        if (((hit_kind) == (2L))) {
                            cx = s0x;
                            cy = s0y;
                            cz = s0z;
                            rad = 0.65;
                        } else {
                            if (((hit_kind) == (3L))) {
                                cx = s1x;
                                cy = s1y;
                                cz = s1z;
                                rad = 0.72;
                            } else {
                                cx = s2x;
                                cy = s2y;
                                cz = s2z;
                                rad = 0.58;
                            }
                        }
                        hx = cam_x + best_t * dx;
                        double hy = cam_y + best_t * dy;
                        hz = cam_z + best_t * dz;
                        java.util.ArrayList<Object> __tuple_6 = ((java.util.ArrayList<Object>)(Object)(normalize((hx - cx) / rad, (hy - cy) / rad, (hz - cz) / rad)));
                        double nx = ((Double)(__tuple_6.get(0)));
                        double ny = ((Double)(__tuple_6.get(1)));
                        double nz = ((Double)(__tuple_6.get(2)));
                        java.util.ArrayList<Object> __tuple_7 = ((java.util.ArrayList<Object>)(Object)(reflect(dx, dy, dz, nx, ny, nz)));
                        double rdx = ((Double)(__tuple_7.get(0)));
                        double rdy = ((Double)(__tuple_7.get(1)));
                        double rdz = ((Double)(__tuple_7.get(2)));
                        java.util.ArrayList<Object> __tuple_8 = ((java.util.ArrayList<Object>)(Object)(refract(dx, dy, dz, nx, ny, nz, 1.0 / 1.45)));
                        double tdx = ((Double)(__tuple_8.get(0)));
                        double tdy = ((Double)(__tuple_8.get(1)));
                        double tdz = ((Double)(__tuple_8.get(2)));
                        java.util.ArrayList<Object> __tuple_9 = ((java.util.ArrayList<Object>)(Object)(sky_color(rdx, rdy, rdz, tphase)));
                        double sr = ((Double)(__tuple_9.get(0)));
                        double sg = ((Double)(__tuple_9.get(1)));
                        double sb = ((Double)(__tuple_9.get(2)));
                        java.util.ArrayList<Object> __tuple_10 = ((java.util.ArrayList<Object>)(Object)(sky_color(tdx, tdy, tdz, tphase + 0.8)));
                        double tr = ((Double)(__tuple_10.get(0)));
                        double tg = ((Double)(__tuple_10.get(1)));
                        double tb = ((Double)(__tuple_10.get(2)));
                        double cosi = Math.max((-(dx * nx + dy * ny + dz * nz)), 0.0);
                        double fr = schlick(cosi, 0.04);
                        r = tr * (1.0 - fr) + sr * fr;
                        g = tg * (1.0 - fr) + sg * fr;
                        b = tb * (1.0 - fr) + sb * fr;
                        lxv = lx - hx;
                        lyv = ly - hy;
                        lzv = lz - hz;
                        java.util.ArrayList<Object> __tuple_11 = ((java.util.ArrayList<Object>)(Object)(normalize(lxv, lyv, lzv)));
                        ldx = ((Double)(__tuple_11.get(0)));
                        ldy = ((Double)(__tuple_11.get(1)));
                        ldz = ((Double)(__tuple_11.get(2)));
                        ndotl = Math.max(nx * ldx + ny * ldy + nz * ldz, 0.0);
                        java.util.ArrayList<Object> __tuple_12 = ((java.util.ArrayList<Object>)(Object)(normalize(ldx - dx, ldy - dy, ldz - dz)));
                        double hvx = ((Double)(__tuple_12.get(0)));
                        double hvy = ((Double)(__tuple_12.get(1)));
                        double hvz = ((Double)(__tuple_12.get(2)));
                        double ndoth = Math.max(nx * hvx + ny * hvy + nz * hvz, 0.0);
                        double spec = ndoth * ndoth;
                        spec = spec * spec;
                        spec = spec * spec;
                        spec = spec * spec;
                        glow = 10.0 / (1.0 + lxv * lxv + lyv * lyv + lzv * lzv);
                        r += 0.2 * ndotl + 0.8 * spec + 0.45 * glow;
                        g += 0.18 * ndotl + 0.6 * spec + 0.35 * glow;
                        b += 0.26 * ndotl + 1.0 * spec + 0.65 * glow;
                        if (((hit_kind) == (2L))) {
                            r *= 0.95;
                            g *= 1.05;
                            b *= 1.1;
                        } else {
                            if (((hit_kind) == (3L))) {
                                r *= 1.08;
                                g *= 0.98;
                                b *= 1.04;
                            } else {
                                r *= 1.02;
                                g *= 1.1;
                                b *= 0.95;
                            }
                        }
                    }
                }
                r = math.sqrt(clamp01(r));
                g = math.sqrt(clamp01(g));
                b = math.sqrt(clamp01(b));
                frame.set((int)((((row_base + px) < 0L) ? (((long)(frame.size())) + (row_base + px)) : (row_base + px))), quantize_332(r, g, b));
            }
        }
        return PyRuntime.__pytra_bytearray(frame);
    }

    public static void run_16_glass_sculpture_chaos() {
        long width = 320L;
        long height = 240L;
        long frames_n = 72L;
        String out_path = "sample/out/16_glass_sculpture_chaos.gif";
        double start = time.perf_counter();
        java.util.ArrayList<java.util.ArrayList<Long>> frames = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long i = 0L; i < frames_n; i += 1L) {
            frames.add(render_frame(width, height, i, frames_n));
        }
        gif.save_gif(out_path, width, height, frames, palette_332(), 6L, 0L);
        double elapsed = time.perf_counter() - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("frames:") + " " + String.valueOf(frames_n));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }
}

File: sample/nim/16_glass_sculpture_chaos.nim
include "py_runtime.nim"

import std/os, std/times, std/tables, std/strutils, std/math

proc clamp01*(v: float): auto =
  if (v < 0.0):
    return 0.0
  if (v > 1.0):
    return 1.0
  return v

proc dot*(ax: float, ay: float, az: float, bx: float, by: float, bz: float): auto =
  return (((ax * bx) + (ay * by)) + (az * bz))

proc length*(x: float, y: float, z: float): auto =
  return sqrt((((x * x) + (y * y)) + (z * z)))

proc normalize*(x: float, y: float, z: float): auto =
  var l = length(x, y, z)
  if (l < 1e-09):
    return (0.0, 0.0, 0.0)
  return ((float(x) / float(l)), (float(y) / float(l)), (float(z) / float(l)))

proc reflect*(ix: float, iy: float, iz: float, nx: float, ny: float, nz: float): auto =
  var d = (dot(ix, iy, iz, nx, ny, nz) * 2.0)
  return ((ix - (d * nx)), (iy - (d * ny)), (iz - (d * nz)))

proc refract*(ix: float, iy: float, iz: float, nx: float, ny: float, nz: float, eta: float): auto =
  var cosi = (-dot(ix, iy, iz, nx, ny, nz))
  var sint2 = ((eta * eta) * (1.0 - (cosi * cosi)))
  if (sint2 > 1.0):
    return reflect(ix, iy, iz, nx, ny, nz)
  var cost = sqrt((1.0 - sint2))
  var k = ((eta * cosi) - cost)
  return (((eta * ix) + (k * nx)), ((eta * iy) + (k * ny)), ((eta * iz) + (k * nz)))

proc schlick*(cos_theta: float, f0: float): auto =
  var m = (1.0 - cos_theta)
  return (f0 + ((1.0 - f0) * ((((m * m) * m) * m) * m)))

proc sky_color*(dx: float, dy: float, dz: float, tphase: float): auto =
  var t = (0.5 * (dy + 1.0))
  var r = (0.06 + (0.2 * t))
  var g = (0.1 + (0.25 * t))
  var b = (0.16 + (0.45 * t))
  var band = (0.5 + (0.5 * math.sin((((8.0 * dx) + (6.0 * dz)) + tphase))))
  r += /* unknown expr Unbox */
  g += /* unknown expr Unbox */
  b += /* unknown expr Unbox */
  return (clamp01(r), clamp01(g), clamp01(b))

proc sphere_intersect*(ox: float, oy: float, oz: float, dx: float, dy: float, dz: float, cx: float, cy: float, cz: float, radius: float): auto =
  var lx = (ox - cx)
  var ly = (oy - cy)
  var lz = (oz - cz)
  var b = (((lx * dx) + (ly * dy)) + (lz * dz))
  var c = ((((lx * lx) + (ly * ly)) + (lz * lz)) - (radius * radius))
  var h = ((b * b) - c)
  if (h < 0.0):
    return (-1.0)
  var s = sqrt(h)
  var t0 = ((-b) - s)
  if (t0 > 0.0001):
    return t0
  var t1 = ((-b) + s)
  if (t1 > 0.0001):
    return t1
  return (-1.0)

proc palette_332*(): auto =
  var p = newSeq[uint8]()
  var v_hoisted_cast_1: float = float(7)
  var v_hoisted_cast_2: float = float(3)
  for i in 0 ..< 256:
    var r = ((i + 5) + 7)
    var g = ((i + 2) + 7)
    var b = (i + 3)
    p[((i * 3) + 0)] = int((float((255 * r)) / float(v_hoisted_cast_1)))
    p[((i * 3) + 1)] = int((float((255 * g)) / float(v_hoisted_cast_1)))
    p[((i * 3) + 2)] = int((float((255 * b)) / float(v_hoisted_cast_2)))
  return bytes(p)

proc quantize_332*(r: float, g: float, b: float): auto =
  var rr = int((clamp01(r) * 255.0))
  var gg = int((clamp01(g) * 255.0))
  var bb = int((clamp01(b) * 255.0))
  return ((((rr + 5) + 5) + ((gg + 5) + 2)) + (bb + 6))

proc render_frame*(width: int, height: int, frame_id: int, frames_n: int): auto =
  var t = (float(frame_id) / float(frames_n))
  var tphase = ((2.0 * math.pi) * t)
  var cam_r = 3.0
  var cam_x = (cam_r * math.cos((tphase * 0.9)))
  var cam_y = (1.1 + (0.25 * math.sin((tphase * 0.6))))
  var cam_z = (cam_r * math.sin((tphase * 0.9)))
  var look_x = 0.0
  var look_y = 0.35
  var look_z = 0.0
  (fwd_x, fwd_y, fwd_z) = normalize((look_x - cam_x), (look_y - cam_y), (look_z - cam_z))
  (right_x, right_y, right_z) = normalize(fwd_z, 0.0, (-fwd_x))
  (up_x, up_y, up_z) = normalize(((right_y * fwd_z) - (right_z * fwd_y)), ((right_z * fwd_x) - (right_x * fwd_z)), ((right_x * fwd_y) - (right_y * fwd_x)))
  var s0x = (0.9 * math.cos((1.3 * tphase)))
  var s0y = (0.15 + (0.35 * math.sin((1.7 * tphase))))
  var s0z = (0.9 * math.sin((1.3 * tphase)))
  var s1x = (1.2 * math.cos(((1.3 * tphase) + 2.094)))
  var s1y = (0.1 + (0.4 * math.sin(((1.1 * tphase) + 0.8))))
  var s1z = (1.2 * math.sin(((1.3 * tphase) + 2.094)))
  var s2x = (1.0 * math.cos(((1.3 * tphase) + 4.188)))
  var s2y = (0.2 + (0.3 * math.sin(((1.5 * tphase) + 1.9))))
  var s2z = (1.0 * math.sin(((1.3 * tphase) + 4.188)))
  var lr = 0.35
  var lx = (2.4 * math.cos((tphase * 1.8)))
  var ly = (1.8 + (0.8 * math.sin((tphase * 1.2))))
  var lz = (2.4 * math.sin((tphase * 1.8)))
  var frame = newSeq[uint8]()
  var aspect = (float(width) / float(height))
  var fov = 1.25
  var v_hoisted_cast_3: float = float(height)
  var v_hoisted_cast_4: float = float(width)
  for py in 0 ..< height:
    var row_base = (py * width)
    var sy = (1.0 - (float((2.0 * (py + 0.5))) / float(v_hoisted_cast_3)))
    for px in 0 ..< width:
      var sx = (((float((2.0 * (px + 0.5))) / float(v_hoisted_cast_4)) - 1.0) * aspect)
      var rx = (fwd_x + (fov * ((sx * right_x) + (sy * up_x))))
      var ry = (fwd_y + (fov * ((sx * right_y) + (sy * up_y))))
      var rz = (fwd_z + (fov * ((sx * right_z) + (sy * up_z))))
      (dx, dy, dz) = normalize(rx, ry, rz)
      var best_t = 1000000000.0
      var hit_kind = 0
      var r = 0.0
      var g = 0.0
      var b = 0.0
      if (dy < (-1e-06)):
        var tf = (float(((-1.2) - cam_y)) / float(dy))
        if py_truthy(((tf > 0.0001) {op} (tf < best_t))):
          best_t = /* unknown expr Unbox */
          hit_kind = 1
      var t0 = sphere_intersect(cam_x, cam_y, cam_z, dx, dy, dz, s0x, s0y, s0z, 0.65)
      if py_truthy(((t0 > 0.0) {op} (t0 < best_t))):
        best_t = t0
        hit_kind = 2
      var t1 = sphere_intersect(cam_x, cam_y, cam_z, dx, dy, dz, s1x, s1y, s1z, 0.72)
      if py_truthy(((t1 > 0.0) {op} (t1 < best_t))):
        best_t = t1
        hit_kind = 3
      var t2 = sphere_intersect(cam_x, cam_y, cam_z, dx, dy, dz, s2x, s2y, s2z, 0.58)
      if py_truthy(((t2 > 0.0) {op} (t2 < best_t))):
        best_t = t2
        hit_kind = 4
      if (hit_kind == 0):
        (r, g, b) = sky_color(dx, dy, dz, tphase)
      elif (hit_kind == 1):
        var hx = (cam_x + (best_t * dx))
        var hz = (cam_z + (best_t * dz))
        var cx = int(math.floor((hx * 2.0)))
        var cz = int(math.floor((hz * 2.0)))
        var checker = /* unknown expr IfExp */
        var base_r = /* unknown expr IfExp */
        var base_g = /* unknown expr IfExp */
        var base_b = /* unknown expr IfExp */
        var lxv = (lx - hx)
        var lyv = (ly - (-1.2))
        var lzv = (lz - hz)
        (ldx, ldy, ldz) = normalize(lxv, lyv, lzv)
        var ndotl = max(ldy, 0.0)
        var ldist2 = (((lxv * lxv) + (lyv * lyv)) + (lzv * lzv))
        var glow = (float(8.0) / float((1.0 + ldist2)))
        r = /* unknown expr Unbox */
        g = /* unknown expr Unbox */
        b = /* unknown expr Unbox */
      else:
        cx = 0.0
        var cy = 0.0
        cz = 0.0
        var rad = 1.0
        if (hit_kind == 2):
          cx = /* unknown expr Unbox */
          cy = /* unknown expr Unbox */
          cz = /* unknown expr Unbox */
          rad = 0.65
        elif (hit_kind == 3):
          cx = /* unknown expr Unbox */
          cy = /* unknown expr Unbox */
          cz = /* unknown expr Unbox */
          rad = 0.72
        else:
          cx = /* unknown expr Unbox */
          cy = /* unknown expr Unbox */
          cz = /* unknown expr Unbox */
          rad = 0.58
        hx = (cam_x + (best_t * dx))
        var hy = (cam_y + (best_t * dy))
        hz = (cam_z + (best_t * dz))
        (nx, ny, nz) = normalize((float((hx - cx)) / float(rad)), (float((hy - cy)) / float(rad)), (float((hz - cz)) / float(rad)))
        (rdx, rdy, rdz) = reflect(dx, dy, dz, nx, ny, nz)
        (tdx, tdy, tdz) = refract(dx, dy, dz, nx, ny, nz, (float(1.0) / float(1.45)))
        (sr, sg, sb) = sky_color(rdx, rdy, rdz, tphase)
        (tr, tg, tb) = sky_color(tdx, tdy, tdz, (tphase + 0.8))
        var cosi = max((-(((dx * nx) + (dy * ny)) + (dz * nz))), 0.0)
        var fr = schlick(cosi, 0.04)
        r = /* unknown expr Unbox */
        g = /* unknown expr Unbox */
        b = /* unknown expr Unbox */
        lxv = (lx - hx)
        lyv = (ly - hy)
        lzv = (lz - hz)
        (ldx, ldy, ldz) = normalize(lxv, lyv, lzv)
        ndotl = max((((nx * ldx) + (ny * ldy)) + (nz * ldz)), 0.0)
        (hvx, hvy, hvz) = normalize((ldx - dx), (ldy - dy), (ldz - dz))
        var ndoth = max((((nx * hvx) + (ny * hvy)) + (nz * hvz)), 0.0)
        var spec = (ndoth * ndoth)
        spec = (spec * spec)
        spec = (spec * spec)
        spec = (spec * spec)
        glow = (float(10.0) / float((((1.0 + (lxv * lxv)) + (lyv * lyv)) + (lzv * lzv))))
        r += (((0.2 * ndotl) + (0.8 * spec)) + (0.45 * glow))
        g += (((0.18 * ndotl) + (0.6 * spec)) + (0.35 * glow))
        b += (((0.26 * ndotl) + (1.0 * spec)) + (0.65 * glow))
        if (hit_kind == 2):
          r *= 0.95
          g *= 1.05
          b *= 1.1
        elif (hit_kind == 3):
          r *= 1.08
          g *= 0.98
          b *= 1.04
        else:
          r *= 1.02
          g *= 1.1
          b *= 0.95
      r = /* unknown expr Unbox */
      g = /* unknown expr Unbox */
      b = /* unknown expr Unbox */
      frame[(row_base + px)] = quantize_332(r, g, b)
  return bytes(frame)

proc run_16_glass_sculpture_chaos*(): auto =
  var width = 320
  var height = 240
  var frames_n = 72
  var out_path = "sample/out/16_glass_sculpture_chaos.gif"
  var start = epochTime()
  var frames: seq[auto] = @[]
  for i in 0 ..< frames_n:
    frames.add(render_frame(width, height, i, frames_n))
  discard save_gif(out_path, width, height, frames, palette_332())
  var elapsed = (epochTime() - start)
  echo "output:", out_path
  echo "frames:", frames_n
  echo "elapsed_sec:", elapsed


if isMainModule:
  run_16_glass_sculpture_chaos()
